multiraft/multiraft_test.go

```diff
@@ -392,6 +392,70 @@ func TestMembershipChange(t *testing.T) {
 		}*/
 }
 
+func TestRemoveLeader(t *testing.T) {
+	defer leaktest.AfterTest(t)
+	stopper := stop.NewStopper()
+	cluster := newTestCluster(nil, 6, stopper, t)
+	defer stopper.Stop()
+
+	// Consume and apply the membership change events.
+	for i := 0; i < 4; i++ {
+		go func(i int) {
+			for {
+				if e, ok := <-cluster.events[i].MembershipChangeCommitted; ok {
+					e.Callback(nil)
+				} else {
+					return
+				}
+			}
+		}(i)
+	}
+
+	// Tick all the clocks in the background to ensure that all the
+	// necessary elections are triggered.
+	// TODO(bdarnell): newTestCluster should have an option to use a
+	// real clock instead of a manual one.
+	stopper.RunWorker(func() {
+		ticker := time.NewTicker(10 * time.Millisecond)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-stopper.ShouldStop():
+				return
+			case <-ticker.C:
+				for _, t := range cluster.tickers {
+					t.Tick()
+				}
+			}
+		}
+	})
+
+	// Create a group with three members.
+	groupID := proto.RangeID(1)
+	cluster.createGroup(groupID, 0, 3)
+
+	// Move the group one node at a time from the first three nodes to
+	// the last three. In the process, we necessarily remove the leader
+	// and trigger at least one new election among the new nodes.
+	for i := 0; i < 3; i++ {
+		log.Infof("adding node %d", i+3)
```

Comment:
`t.Logf` messages are buffered by the testing framework and printed at the end of the test, which means that they can't be correlated with any of our other log spam. I don't think it ever makes sense for us to use `t.Logf` from a package that already uses our `log` package.